### PR TITLE
feat: include typescript support

### LINKED ISSUE #346: Add TypeScript support
I use `react-intl-tel-input` for a variety of projects over the past 3 years but recently, I've included `typescript` because the latest project I'm working on is quite big. I want to have a reasonable amount of strictly-typed control over my components which this library doesn't have at the time of writing.

## Expected Behavior

This library should already have type definitions out of the box.

## Current Behavior

Without any type definitions, the TypeScript compiler will complain. You need to get definitions from DefinitelyTyped (which doesn't exist yet) or declare the types yourself in every new project you write (inconvenient, of course).

image

## Possible Solution

I've written drop-in support that you can check out [in this gist](https://gist.github.com/andrewsantarin/cb45e6370c50aaabb374aaca34ae100f). It was built to match the prop types of 7.x, but I imagine that 8.x hasn't changed anything, so those type definitions wouldn't be far off from the real thing:

https://github.com/patw0929/react-intl-tel-input/blob/d769a1e9a8b5b921b1f9bef74e8b6e75ceea7d73/src/components/IntlTelInput.js#L1324-L1401

If there are more modules exported by this library, then we should write type definitions for them, too, for full support.

## Steps to Reproduce

1. Create a React TypeScript sandbox: https://codesandbox.io/s/create-react-app-typescript-enhvd (my template)
2. Insert the Code below to the top of App.tsx.
3. Expect the results shown in the Current Behavior.

Code:

```ts
import IntlTelInput from 'react-intl-tel-input';
```

Live Example:
https://codesandbox.io/s/serene-leakey-b8uph

This is with the type definitions applied.
To simulate what it looks like **without** type definitions, simply remove `index.d.ts` in the example.

## Environment

* Version:
  
* Browser:

### PR DESCRIPTION
**Pull request by andrewsantarin**
This PR adds type definition modules so that React TypeScript projects using this library do not need to introduce a drop-in module augmentation of their own, such as in [this example gist](https://gist.github.com/andrewsantarin/cb45e6370c50aaabb374aaca34ae100f).

Resolves #346.

## Description
- Added `.d.ts` modules **for each file exported** by this library rather than the main export only (which [the gist](https://gist.github.com/andrewsantarin/cb45e6370c50aaabb374aaca34ae100f) does). The type definitions are up to date with the latest APIs.
- Added a script for outputting the type definition modules to the production build.
  - Used `gulp` to perform simply one-to-one copying of the `.d.ts` modules from `src` to `dist`.
- Added tests in `src/components/__tests__` alongside the existing `.js` tests:
  - Used `tsc` (using `tsconfig.test.json`) to verify that all imports are working correctly.
  - Used `tsd` (using `tsconfig.json` and the `"tsd"` override field in `package.json`) to test the correctness of functions which are overloadable or generic or both.
- Added instructions in the `README.md` file on:
  - How to use this library in a TypeScript project.
  - How to test the TypeScript modules when making contributions to this project.

## Screenshots (if appropriate):
Not a screenshot, but a demo in [CodeSandbox](https://codesandbox.io/s/react-intl-tel-input-typescript-support-pre-merge-higuz) of the type definition modules in action with a pre-merge copy of this library: [`@andrewsantarin/react-intl-tel-input`](https://www.npmjs.com/package/@andrewsantarin/react-intl-tel-input)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have used ESLint & Prettier to follow the code style of this project.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by andrewsantarin** [comment]
Sorry, it took me a while to get back to you, @mcataford! I had a lot of internal debate these past few days on how to make everything work. It's not my first GitHub rodeo, but it's my first attempt at something bigger than adding something as simple as a `index.d.ts` file. It's definitely a learning process for me and I look forward to watching it pay off!

Here some decisions I've made in organizing this PR:
- I didn't rewrite the entire library in TypeScript. Why?
  - Those who understand JavaScript only will have to (re-)learn TypeScript. If they're not up for it, development of this project will slow down.
  - There is a rather large API surface area right now, **especially** with the main export (a whopping 50 class methods, 22 class properties, 34 React props, 12 React states), of which some have types which are very difficult to infer while reading through the code. To rewrite all of this into `.tsx` would be a huge undertaking which I'm not willing to commit to. I'd guarantee you it'd take me weeks plus a lot of back-n'-forth Q&A.
  - The build process may have to change. As a novice author, I'm very unfamiliar with build processes, (`rollup` is all I know and it's very limited). This will also take a lot of time **and** help from seasoned TypeScript authors.
- I've added `gulp` (build) and `tsd` (test) as development dependencies.
  - I would've used `typescript`'s compiler instead of `gulp` if I could help it, but even after a lot of hours researching and trying, I could never get it to export the type definition modules correctly. A proper TypeScript output would have: `class IntlTelInput extends React.Component<IntlTelInputProps, IntlTelInputState> {}` but what I get is `class IntlTelInput extends React.Component<any, any>`. If there's a way to do this correctly without relying on `gulp`, I'm all ears.
  - Checking for correctness of some of the more sophisticated types is where `tsd` comes in. `dtslint` is an alternative. [There's a reason to verify that the types work as intended.](https://www.youtube.com/watch?v=nygcFEwOG8w)
- I've opted to go with `.d.ts` side-by-side with `.js` ([example from `semantic-ui-react`](https://github.com/Semantic-Org/Semantic-UI-React/tree/master/src/elements/Button)). 
  - There's [the other camp](https://moduscreate.com/blog/publishing-typescript-modules-npm/) which suggests a `typings` folder with one-to-one parity with the `src` folder but I didn't like that it's not immediately obvious to me which files **don't** have a `.d.ts` counterpart.
- [`underscore.deferred`](https://github.com/wookiehangover/underscore.deferred), which `IntlTelIput` uses for all 3 of its deferred properties, **doesn't** have a type definition module. The closest counterpart is `jquery`'s Deferred object, which its author claims to achieve parity with. I used `@types/jquery` (just the type module, not the actual library) to infer the type. To be sure, I made a comparison of instances methods between both modules by logging them both in the browser console.
- ESLint & Prettier worked for the `.js` files, but for some reason, it doesn't do anything to the `.d.ts` counterparts. I'm not sure how to solve this. Yes, I did try `eslint src *.d.ts`. But I got nothing, so I did the linting by hand.

I hope these notes will serve as food for thought for the project.

File: package.json [hunk 1]
```diff
@@ -108,13 +120,19 @@
...
     "webpack": "^4.27.1"
   },
   "scripts": {
     "prebuild": "yarn run clean",
-    "build": "BABEL_ENV=production babel src -d dist && cp -r ./src/*.png ./dist && node-sass ./src/intlTelInput.scss ./dist/main.css",
+    "build": "yarn build:js && yarn build:png && yarn build:css && yarn build:ts",
+    "build:js": "cross-env BABEL_ENV=production babel src -d dist",
+    "build:png": "cp -r ./src/*.png ./dist",
+    "build:css": "node-sass ./src/intlTelInput.scss ./dist/main.css",
+    "build:ts": "gulp build:ts",
```
Comment (andrewsantarin) [inline package.json:135]: I've split this up because it was too long to read in my screen, especially now with the `gulp` build added.

**Comment by mcataford** [comment]
Wow @andrewsantarin! That's quite a piece of work! Since it touches a whole lot, I'll shelf ongoing branches on my end since they would probably conflict with this. I'll allocate some time ASAP to read through and review! Thanks for the time you've put in this!

**Comment by andrewsantarin** [comment]
> I'll shelf ongoing branches on my end since they would probably conflict with this.

@mcataford Hmmm... I doubt any of those other branches will (**none** of my work changes `.js` files) as long as the APIs don't change. If they will, then I believe a simple type definition update would be in order.

But it's your call. I'm sure you're already on top of things and I'd rather not upset that!

**Comment by andrewsantarin** [comment]
@mcataford @patw0929 It has been 8 months since the last major update. When can we see this merged into `master`?

**Comment by mcataford** [comment]
Hi @andrewsantarin! Sorry for the ultralong wait! Ready when you are -- I spent some time earlier this week fixing up CI and devDependency problems that prevented new builds on the library, but now that this is done, we should be good to move ahead with more changes! Let's bring it up to date with `master` and have CI do its thing, then we can get it in! :rocket:

**Comment by andrewsantarin** [comment]
@mcataford I see that we have some merge conflicts here, but sure, I'd be happy to oblige. Could you wait till weekend?

**Comment by mcataford** [comment]
Absolutely! No rush! Enabling TS is the next big thing on my mind. I don't anticipate a whole lot of movement that would create more conflicts (the big conflicting bit probably being the Yarn > Yarn Berry move).

I'll keep an eye out for updates on this PR whenever you are ready and have time to check it out!

**Comment by andrewsantarin** [comment]
@mcataford I'll go through this PR again and will add you as a reviewer as soon as it's ready. TypeScript has gone up one major version since my last update. I doubt anything would break but it'd be nice to keep the dependencies green.

**Comment by andrewsantarin** [comment]
Closing this. Will submit a cleaner PR.

**Pull request closed without merging**